### PR TITLE
Use Ansible for building wheels and provisioning docker images.

### PR DESCRIPTION
**Pull request by mateuszlewko**
The first version of Ansible playbook that can build PyTorch and XLA wheels and provision docker image (`release` stage). 
The playbook can install different pip / apt packages and set environment variable depending on the specified stage, architecture and accelerator.

The playbook we'll be later invoked in a dockerfile, so the config structure or Ansible roles may change a little in the future.

### CODE REVIEW

**Comment by JackCaoG** [comment]
Thanks @mateuszlewko ! The branch cut for pytorch 2.0 will be end of this week so I would hold on review this pr until we cut the branch.

File: docker/experimental/ansible/README.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# Ansible playbook
+
+This ansible playbook will perform the following actions on the localhost:
+  * install required pip and apt packages, depending on the specified stage,
+    architecture and accelerator (see [apt.yaml](config/apt.yaml) and
+    [pip.yaml](config/pip.yaml)).
+  * fetch bazel (version configured in [vars.yaml](config/vars.yaml)),
+  * fetch PyTorch and XLA sources at master (or specific revisions,
+    see role `fetch_srcs` in [playbook.yaml]).
```
Comment (stgpetrovic) [inline docker/experimental/ansible/README.md:9]: I think you need [text](url) format here too.

File: docker/experimental/ansible/config/apt.yaml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+      - libopenblas-dev
+      - ninja-build
+      - procps
+      - python3-pip
+      - rename
+      - vim
+      - wget
+
+    build_cuda:
+      - cuda-libraries-11-2
```
Comment (stgpetrovic) [inline docker/experimental/ansible/config/apt.yaml:19]: Recently [bumped](https://github.com/pytorch/xla/commit/a497db46e7fb7b58be7584aed69ac7477428b3c6) to 11.7, but in the circleci, not in docker. Hm.

@JackCaoG what's the source of truth here?

Comment (JackCaoG) [reply docker/experimental/ansible/config/apt.yaml:19]: @yeounoh

File: docker/experimental/ansible/config/pip.yaml [hunk 3]
```diff
@@ -0,0 +1,52 @@
+# Contains lists of pip packages for each stage (build|release) and arch or accelerator.
```
[the remark this answers is not in the document]
Comment (will-cromar) [reply docker/experimental/ansible/config/pip.yaml:1]: I removed conda from the experimental build to simplify the process and integrate better with Bazel in the future, which has support for `pip` packages

File: docker/experimental/ansible/roles/build_srcs/tasks/main.yaml [hunk 4]
```diff
@@ -0,0 +1,46 @@
+- name: Build PyTorch
```
[the remark this answers is not in the document]
Comment (will-cromar) [reply docker/experimental/ansible/roles/build_srcs/tasks/main.yaml:1]: In general, we shouldn't ship a docker image that's used for builds if we're concerned about image size. We should move the compiled packages into a clean image without the build dependencies to reduce the footprint

File: docker/experimental/ansible/roles/fetch_srcs/defaults/main.yaml [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# See https://docs.ansible.com/ansible/latest/collections/ansible/builtin/git_module.html#parameter-version
+pytorch_git_rev: HEAD
```
[the remark this answers is not in the document]
Comment (will-cromar) [reply docker/experimental/ansible/roles/fetch_srcs/defaults/main.yaml:2]: Is there a green branch that we can use? I've been using HEAD this whole time, and I assumed others were as well

File: docker/experimental/ansible/config/apt.yaml (hunk 2 above)
Comment (will-cromar) [reply docker/experimental/ansible/config/apt.yaml:19]: Right now, the builds are very fragmented, so the release images haven't been updated. If possible, we should use 11.7 or 11.8 going forward. If the change here isn't trivial, then we can push the work to a follow up PR

**Review by will-cromar — COMMENTED** [review]
Thanks! This is going to be a huge improvement to the constellation of bash scripts we are using today. I agree that we should hold off on merging this until after the release cut

File: docker/experimental/ansible/roles/fetch_srcs/defaults/main.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (steventk-g) [reply docker/experimental/ansible/roles/fetch_srcs/defaults/main.yaml:2]: There isn't as of now, I guess this is out of scope for the PR. I do wish we had a concept of "the most recent upstream commit that has been validated to build with XLA" that we could fetch for some builds. On the other hand, this way ensures we are keeping up with the upstream HEAD.

[the remark this answers is not in the document]
Comment (stgpetrovic) [reply docker/experimental/ansible/roles/fetch_srcs/defaults/main.yaml:2]: I think this is something we should do soon.

There are ways of automatically picking the next candidate (based on some signal) and creating a PR that bumps the dependency to a newer version and submitting it automatically if the tests pass. We will look into that soon, validate the approach and we'll see if we can do this, but II'm confident that we can get both TF (+openxla/tsl) and PT on tihs schedule.

File: docker/experimental/ansible/roles/build_srcs/tasks/main.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (mateuszlewko) [reply docker/experimental/ansible/roles/build_srcs/tasks/main.yaml:1]: > We should move the compiled packages into a clean image without the build dependencies to reduce the footprint
Yes, I'll make sure that happens in the dockerfile used for building the release image.

File: docker/experimental/ansible/config/apt.yaml (hunk 2 above)
Comment (mateuszlewko) [reply docker/experimental/ansible/config/apt.yaml]: Done

**Comment by JackCaoG** [comment]
Can we rebase this pr? CI failure is unrelated.

**Review by will-cromar — APPROVED** [review]
Thanks!

**Pull request merged**